### PR TITLE
ISSUE-2773: Add db ledgers index rebuild op

### PR DESCRIPTION
**Pull request by Vanlightly**
### Motivation

File corruption of the ledgers index (DbLedgerStorage) has happened in production and we currently have no good way of resolving it. There exists a locations index rebuild command, this issue describes a new command targeted at the ledgers index.

This command should only be run when either the bookie is offline, or when in readonly mode to avoid data loss. Container based environments may not be able to make the bookie go offline while also allowing the shell or bkctl to perform the operation.

The db ledgers index rebuild op does the following:

- scans all journal and entry log files to discover all ledger ids currently stored.
- builds a new index where each ledger is fenced and has an empty master key set.
- If all went well, replaces the original index with the new one

The bookie should then be restarted again (in normal mode) to load the rebuilt index. 

Notable stuff:

The reason for setting fencing each ledger is that there is no safe way of setting the fenced status based on metadata while the bookie is running. If the command is run when the bookie is in readonly mode, it can still serve fence requests and any fencing that occurs while the rebuild is occurring will be lost, which breaks the data safety guarantees of the BookKeeper protocol. Given also that the bookie should at least be in readonly mode (else offline), it is likely a member of the current ensemble of few ledgers so the impact should be minimal.
The reason for setting an empty master key is that firstly, an empty master key simply gets overwritten, so cannot cause an IllegalOpException. Secondly, if we use the password stored in metadata, then we need to be sure to use exactly the same digest algorithm as the client when it creates the ledgerKey, else the bookie will start failing all writes. This could potentially cause a problem in the future if the way the ledgerKey is generated changes (old clients would be incompatible after a ledgers index rebuild).

### Changes

Adds two new commands. "rebuild-db-ledgers-index" rebuilds the db ledgers index by scanning the journal and entry log files. "check-db-ledgers-index" performs a read scan of the db ledgers index, this can be used when corruption is suspected.

Additionally:
- a test of the new rebuild op
- makes the Journal. listJournalIds() method public so that the command can iterate over the journal files.
- rewords some printed statements in the LocationsIndexRebuildOp to make it clear which index it is rebuilding (given that there are two indexes in the DbLedgerStorage).

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java:120]: why even move the original at the start? it creates a lot of surface to fail, and then you have to unwind. Instead you could create the new index, and if successful, move old and move in new, which have fewer ops that can fail.

Comment (Vanlightly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java]: Good point, I was copying from the LocationsIndexRebuildOp which does it that way. I have updated this rebuild to swap them at the end once a new index has been generated.

**Comment by Vanlightly** [comment]
@ivankelly @eolivelli any chance of a review?

**Review by eolivelli — APPROVED** [review]
Very good

+1

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
great work!

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java
Comment (Vanlightly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java:121]: If the first move fails, then the index should be in its original or back-up location. If the second move fails, the rebuilt index will still be in its temp location. Rather than try to auto fix after an error here, I'll add better logging to help the administrator fix it themselves. Seems like a highly unlikely scenario for a move to fail.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java
Comment (Vanlightly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java:93]: Ok

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexCheckOp.java
Comment (Vanlightly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexCheckOp.java:83]: Ok

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java
Comment (Vanlightly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java:125]: I'll let the admin decide what to do here. I have added better logging of the move ops to help a human decide how to fix it.

**Comment by hsaputra** [comment]
rerun failure checks

**Comment by hsaputra** [comment]
Will merge this EOD PST if no more comments. Thanks

**Pull request merged**
